usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateRouteColorAndTextContrast.java

```diff
@@ -0,0 +1,101 @@
+
+    /**
+     * @param dataRepo   a repository storing the data of a GTFS dataset
+     * @param resultRepo a repository storing information about the validation process
+     */
+    public ValidateRouteColorAndTextContrast(final GtfsDataRepository dataRepo,
+                                             final ValidationResultRepository resultRepo) {
+        this.dataRepo = dataRepo;
+        this.resultRepo = resultRepo;
+    }
+
+    /**
+     * @param color an hexadecimal string representing a color
+     * @return the resulting luminance for the color
+     */
+    private double luminance(String color) {
+        int r = Integer.parseInt(color.substring(0, 2), 16);
+        int g = Integer.parseInt(color.substring(2, 4), 16);
+        int b = Integer.parseInt(color.substring(4, 6), 16);
+
+        return (0.2126 * r + 0.7152 * g + 0.0722 * b);
+    }
+
+    /**
+     * @param routeColor an hexadecimal string representing a Route color
+     * @param textColor  an hexadecimal string representing a Route text color
+     * @return the contrast ratio between the 2 colors
+     */
+    private double contrast(String routeColor, String textColor) {
+        double routeLuminance = luminance(routeColor);
+        double textLuminance = luminance(textColor);
+        double brightest = Math.max(routeLuminance, textLuminance);
+        double darkest = Math.min(routeLuminance, textLuminance);
+
+        return (brightest + 0.05) / (darkest + 0.05);
+    }
+
+    /**
+     * Verifies if 2 colors are contrasting according to W3C recommendations
+     * Ref. https://www.w3.org/TR/WCAG20-TECHS/G17.html#G17-procedure
+     *
+     * @param routeColor an hexadecimal string representing a Route color
+     * @param textColor  an hexadecimal string representing a Route text color
+     * @return true if contrast ratio between the 2 colors is higher than 4.5, false if not.
+     */
+    private boolean areContrasting(String routeColor, String textColor) {
+        return contrast(routeColor, textColor) >= 4.5;
+    }
+
+    /**
+     * Use case execution method: checks if Route description is the same as Route long and short names
+     * for every Routes in a {@link GtfsDataRepository}. A new notice is generated each time this condition is true.
+     * This notice is then added to the {@link ValidationResultRepository} provided in the constructor.
+     *
+     * @return a list of notices generated each time a Route description equals the Route long or short name.
+     */
+    public void execute() {
+        Collection<Route> routes = dataRepo.getRouteAll();
+        routes.stream()
+                .filter(route -> route.getRouteColor() != null &&
+                        route.getRouteTextColor() != null &&
+                        !areContrasting(route.getRouteColor(), route.getRouteTextColor()))
+                .forEach(route -> resultRepo.addNotice(new RouteColorAndTextContrastNotice("route.txt", route.getRouteId())));
```

Comment:
As discussed, find a way to make the notice more explicit ie: "Ratio should be >= 4.5 but was 2.6"

Reply:
The Python validator has some good comments on why certain thresholds are used, which we should include in our code too:
https://github.com/google/transitfeed/blob/master/transitfeed/route.py#L231

The error message is also very clear here and indicates common problems as to why there isn't contrast:
https://github.com/google/transitfeed/blob/master/transitfeed/route.py#L231

If a GTFS dataset fails this rule, the next natural question an agency would ask is "well, if these colors aren't contrasting, what's the nearest contrasting color I can use?" I remember actually dealing with this issue (IIRC when picking background/foreground colors on Android) and it took me a while to hunt down tools to actually measure contrast for a given color. It would be great if we could point to a web tool that allows you to input hex colors and get a contrast measurement back that agencies could use to fix their colors.

Reply:
More information was added to this notice, including error text from the Python validator, and a link to the contrast ratio formula.